src/com/google/javascript/jscomp/RewriteJsonToModule.java

```diff
@@ -0,0 +1,150 @@
+
+  private final Map<String, String> packageJsonMainEntries;
+  private final Compiler compiler;
+
+  /**
+   * Creates a new RewriteJsonToModule instance which can be used to
+   * rewrite JSON files to modules.
+   *
+   * @param compiler The compiler
+   */
+  public RewriteJsonToModule(Compiler compiler) {
+    this.compiler = compiler;
+    this.packageJsonMainEntries = new HashMap<>();
+  }
+
+  public ImmutableMap<String, String> getPackageJsonMainEntries() {
+    return ImmutableMap.copyOf(packageJsonMainEntries);
+  }
+
+  /**
+   * Module rewriting is done a on per-file basis prior to main compilation.
+   * The root node for each file is a SCRIPT - not the typical jsRoot of other passes.
+   */
+  @Override
+  public void process(Node externs, Node root) {
+    Preconditions.checkState(root.isScript());
+    NodeTraversal.traverseEs6(compiler, root, this);
+  }
+
+  @Override
+  public void visit(NodeTraversal t, Node n, Node parent) {
+    switch (n.getToken()) {
+      case SCRIPT:
+        if (n.getChildCount() != 1) {
+          compiler.report(t.makeError(n, JSON_UNEXPECTED_TOKEN));
+        } else {
+          visitScript(t, n, parent);
+        }
+        return;
+
+      case OBJECTLIT:
+      case ARRAYLIT:
+      case NUMBER:
+      case TRUE:
+      case FALSE:
+      case NULL:
+      case STRING:
+        break;
+
+      case STRING_KEY:
+        if (!n.isQuotedString() || n.getChildCount() != 1) {
+          compiler.report(t.makeError(n, JSON_UNEXPECTED_TOKEN));
+        }
+        break;
+
+      case EXPR_RESULT:
+        if (!parent.isScript()) {
+          compiler.report(t.makeError(n, JSON_UNEXPECTED_TOKEN));
+        }
+        break;
+
+      default:
+        compiler.report(t.makeError(n, JSON_UNEXPECTED_TOKEN));
+        break;
+    }
+
+    if (n.getLineno() == 1) {
+      // We wrapped the expression in parens so our first-line columns are off by one.
+      // We need to correct for this.
+      n.setCharno(n.getCharno() - 1);
+      compiler.reportCodeChange();
+    }
+  }
+
+  /**
+   * For script nodes of JSON objects, add a module variable assignment
+   * so the result is exported.
+   *
+   * If the file path ends with "/package.json", look for a "main"
+   * key in the object literal and track it as a module alias.
+   */
+  private void visitScript(NodeTraversal t, Node n, Node parent) {
+    if (n.getChildCount() != 1 || !n.getFirstChild().isExprResult()) {
+      compiler.report(t.makeError(n, JSON_UNEXPECTED_TOKEN));
+      return;
+    }
+
+    Node jsonObject = n.getFirstFirstChild().detach();
+    n.removeFirstChild();
+
+    String moduleName = t.getInput().getPath().toModuleName();
+
+    n.addChildToFront(IR.var(IR.name(moduleName).useSourceInfoFrom(jsonObject),
+        jsonObject).useSourceInfoFrom(jsonObject));
+
+    n.addChildToFront(
+        IR.exprResult(
+            IR.call(IR.getprop(IR.name("goog"), IR.string("provide")), IR.string(moduleName)))
```

Comment:
This is for use with Node.js, right?
Shouldn't we be rewriting this to look like a non-JSON Node.js module then, rather than a goog.provide()?
Perhaps Node.js modules get rewritten to look like goog.provide() anyway?
Even if that's so, isn't it safer to rewrite this as a Node.js module & allow the later pass to rewrite it again as a goog.provide(). Otherwise, we're duplicating the logic here & risking inconsistent behavior.